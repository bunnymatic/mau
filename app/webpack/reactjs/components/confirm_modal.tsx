import { noop } from "@js/app/helpers";
import { buttonStyleAttrs, MauButton } from "@reactjs/components/mau_button";
import { MauModal, setAppElement } from "@reactjs/components/mau_modal";
import { useModalState } from "@reactjs/hooks";
import * as types from "@reactjs/types";
import cx from "classnames";
import React, { type FC, type ReactNode } from "react";

type ConfirmModalHandler = (success: boolean) => void;

export enum ConfirmModalVariants {
  normal = "normal",
  large = "large",
}

interface ConfirmModalProps {
  id?: string;
  buttonText: string;
  buttonStyle?: types.MauButtonStyle;
  handleConfirm?: ConfirmModalHandler;
  containerClass?: string;
  variant?: "normal" | "large";
  children?: ReactNode;
}

export const ConfirmModal: FC<ConfirmModalProps> = ({
  id,
  buttonStyle,
  buttonText,
  handleConfirm,
  containerClass,
  variant = ConfirmModalVariants.normal,
  children,
}) => {
  const { isOpen, open, close } = useModalState();
  setAppElement("body");

  const submit = (answer: boolean) => (_event: MouseEvent) => {
    (handleConfirm ?? noop)(answer);
    close();
  };

  return (
    <>
      <div id={id} className="mau-modal confirm-modal__trigger">
        <MauButton
          {...buttonStyleAttrs(buttonStyle)}
          onClick={(ev) => {
            ev.preventDefault();
            open();
          }}
        >
          {buttonText}
        </MauButton>
      </div>
      <MauModal isOpen={isOpen}>
        <div
          className={cx("confirm-modal__content", {
            [containerClass]: Boolean(containerClass),
          })}
        >
          <div
            className={cx("confirm-modal__content--main", {
              "confirm-modal__content--main--large":
                variant === ConfirmModalVariants.large,
            })}
          >
            {children}
          </div>
          <div className="confirm-modal__actions">
            <div className="confirm-modal__action-item">
              <MauButton primary onClick={submit(true)}>
                Yes
              </MauButton>
            </div>
            <div className="confirm-modal__action-item">
              <MauButton secondary onClick={submit(false)}>
                No
              </MauButton>
            </div>
          </div>
        </div>
      </MauModal>
    </>
  );
};
