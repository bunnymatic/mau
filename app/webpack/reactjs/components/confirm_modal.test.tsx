import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import React from "react";

import { ConfirmModal } from "./confirm_modal";

describe("ConfirmModal", () => {
  it("works with button text", () => {
    const { container } = render(
      <ConfirmModal id="my-modal" buttonText="click here to confirm">
        this is the confirm message
      </ConfirmModal>
    );
    expect(container).toMatchSnapshot();
  });

  it("works with button text and style", () => {
    const { container } = render(
      <ConfirmModal
        id="my-modal"
        buttonText="click here to confirm"
        buttonStyle="primary"
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("when i click the button the modal opens", async () => {
    act(() => {
      render(
        <ConfirmModal
          id="my-modal"
          buttonText="click here to confirm"
          buttonStyle="primary"
        >
          this is the confirm message
        </ConfirmModal>
      );
    });
    await waitFor(() => {
      expect(
        screen.queryByRole("button", { name: "click here to confirm" })
      ).toBeInTheDocument();
    });
    act(() => {
      const confirm = screen.queryByRole("button", {
        name: "click here to confirm",
      });
      fireEvent.click(confirm);
    });
    await waitFor(() => {
      expect(
        screen.queryByText("this is the confirm message")
      ).toBeInTheDocument();
    });
    act(() => {
      const yes = screen.queryByText("Yes")?.closest("button");
      fireEvent.click(yes);
    });
  });

  it("when i click yes and the button the modal opens", () => {
    const mockConfirm = jest.fn();
    render(
      <ConfirmModal
        id="my-modal"
        buttonText="click here to confirm"
        buttonStyle="primary"
        handleConfirm={mockConfirm}
      >
        this is the confirm message
      </ConfirmModal>
    );
    expect(
      screen.queryByRole("button", { name: "click here to confirm" })
    ).toBeInTheDocument();
    let confirm;
    confirm = screen.queryByRole("button", { name: "click here to confirm" });
    fireEvent.click(confirm);
    expect(
      screen.queryByText("this is the confirm message")
    ).toBeInTheDocument();
    const yes = screen.queryByText("Yes")?.closest("button");
    fireEvent.click(yes);
    expect(mockConfirm).toHaveBeenCalledWith(true);

    confirm = screen.queryByRole("button", { name: "click here to confirm" });
    fireEvent.click(confirm);
    expect(
      screen.queryByText("this is the confirm message")
    ).toBeInTheDocument();
    const no = screen.queryByText("No")?.closest("button");
    fireEvent.click(no);
    expect(mockConfirm).toHaveBeenCalledWith(false);
  });
});
