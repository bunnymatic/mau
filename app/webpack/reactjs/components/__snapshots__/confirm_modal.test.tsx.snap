// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmModal works with button text 1`] = `
<div>
  <div
    class="mau-modal confirm-modal__trigger"
    id="my-modal"
  >
    <button
      class="pure-button"
      role="button"
    >
      click here to confirm
    </button>
  </div>
</div>
`;

exports[`ConfirmModal works with button text and style 1`] = `
<div>
  <div
    class="mau-modal confirm-modal__trigger"
    id="my-modal"
  >
    <button
      class="pure-button button-primary"
      role="button"
    >
      click here to confirm
    </button>
  </div>
</div>
`;
